File: server/src/Models/activityModel.js
import mongoose from "mongoose";

const subActivitySchema = new mongoose.Schema({
  subactivity: {
    type: String,
    required: true,
  },
  haveIndicative: {
    type: Boolean,
    default: false,
    required: true,
  },
});

const activitySchema = new mongoose.Schema({
  area: {
    type: String,
    required: true,
    unique: true,
  },
  activities: {
    type: [
      {
        activity: {
          type: String,
          required: true,
        },
        subactivities: {
          type: [subActivitySchema], 
          required: true,
          default: [],
        },
      },
    ],
    required: true,
  },
});

activitySchema.statics.isUnique = async function (area) {
  try {
    if (!area) throw new Error("Dados inválidos");

    const areaInDatabase = await this.findOne({ area });

    if (areaInDatabase) return false;
    return true;
  } catch (err) {
    console.log("Erro ao verificar dados", err.message);
    return false;
  }
};

export default mongoose.model("Activity", activitySchema);

File: server/src/models/activityModel.go
package models

import (
	"go.mongodb.org/mongo-driver/bson/primitive"
)

type SubActivity struct {
	SubActivity    string `bson:"subactivity" json:"subactivity" validate:"required"`
	HaveIndicative bool   `bson:"haveIndicative,omitempty" json:"haveIndicative"`
}

type ActivitySchema struct {
	Activity      string        `bson:"activity" json:"activity" validate:"required"`
	SubActivities []SubActivity `bson:"subactivities,omitempty" json:"subactivities" validate:"required"`
}

type Activity struct {
	ID         primitive.ObjectID `bson:"_id,omitempty" json:"id"`
	Area       string             `bson:"area" json:"area" validate:"required"`
	Activities []ActivitySchema   `bson:"activities" json:"activities" validate:"required"`
}
